Adding a spec template (#15)
microsoft/fluent-ui-react-staging: specs/ComponentDesignGuidance.md

# Props conformance for components

This document outlines the basic guidance for component design and props naming.

## Basic expectations for every component

> Every component receives an "as".

Reviewer: Missing an - at the start for proper indent as well.

- The component can always provide `id`, `className`, `style`, `aria-*` and `data-*` attributes on the container of a component.

- If the component represents a standard HTML element (e.g. Button), we should mix in the valid root props using the `getNativeProps`
  utility so that we can let the user use native functionality as expected.

- If the component hosts clickable elements, there must be a way to inject `data-*` attributes onto the clickable components. This enables test automation to be built downstream.

- Interactable elements must be accessible via keyboard, mouse, and touch. Nothing can be ONLY interactable by mouse.

> ## Build many smaller components and compose them together.

Reviewer: I think that this also includes how we should build smaller separate state functions and be able to compose that as well.

Often we want to build something complex, like a CommandBar. We see a picture of what we want and we build one giant component that does
this. Then we find other scenarios that have overlaps. The CommandBar contains a SearchBox, a set of left side collapsable links and right
side links. We may find we have other cases that just want a set of links, without the extra overhead. We may also find cases where we just
want a single command bar item with a dropdown menu.

This is an example where instead of building a single large component, we should build more atomic building blocks that are puzzled
together. While it may be more effort to think in smaller blocks, it makes the code easier to reuse and often simpler to maintain.

## Use a .types.ts file to extract out the public contracts that should be supported and documented.

> A `.types` file contains all of the interface contracts that the user should know about to use the component. It is the "contract" for the component. When we evaluate semantic versioning, we look through the changes at `.types` files to determine if the change is a major, minor, or patch.

Reviewer 2: ```suggestion
A `.types` file contains all of the component's user facing interfaces. It is the "contract" for the component. When we evaluate semantic versioning, we look through the changes at `.types` files to determine if the change is a major, minor, or patch.
```

The `.types` files are also auto documented. All JSDoc comments will be extracted and shown on the demo site for documentation.

When your component exposes public methods/properties, define an interface for the component in the `.types` file and implement the interface.
The auto documentation will interpret the I{Component} interface as the class contract and show it in the documentation as the class
definition.

## Imperative API strategies

In rare cases, a component may need to offer an imperative api surface. The most common scenario is to ask a component to focus, but other scenarios may be surfaces such as scrolling to a particular item index for a virtualized list.

> We do not want the user to access private content, so we lock imperative API surfaces down to a `componentRef` reference.

Reviewer 3: Maybe something like:

```suggestion
We do not want the user to access the DOM properties of the element, so we lock imperative API surfaces down to a `componentRef` reference.
```

| buttonDisabled | disabled          | If the subject IS the component, don't give it a subject name.                     |

### Event callbacks

Event callbacks should be prefixed with `on`. If it is necessary to target a subject, the subject should be in between `on` and the event
name. Always use present tense verbs instead of past tense.

| BAD                      | GOOD         | Notes |
| ------------------------ | ------------ | ----- |
| mouseClick               | onClick      |
| onClicked                | onClick      |
| inputClick, onClickInput | onInputClick |

It should also be noted that for event callbacks, the ORDER of parameters should be consistent with the precedence if one exists.

Example:

| BAD                                                  | GOOD                                                       |
| ---------------------------------------------------- | ---------------------------------------------------------- |
| `onClick: (item: any, ev: React.MouseEvent) => void` | `onClick: (ev?: React.MouseEvent<{}>, item?: any) => void` |

### Custom render methods

> In cases where you wish to give the user a way to render a custom version of a thing, follow the convention `onRenderSubject`. It should

Reviewer: I thought that we were getting rid of these entirely with Slots and or the As property?

also use the `IRenderFunction` interface defined in common, which takes in `props` and a `defaultRender` function and returns a JSX element:

BAD:
`renderItem: (item: any, foo: any, bar; any) => JSX.Element`

GOOD:
`onRenderItem: IRenderFunction<IItemProps>`

> ## Use arrow function properties to avoid ALL binds in templates

Reviewer 2: also, arrow functions assume we're using classes. Do we have a hard recommendation on classes vs functions?
